File: apps/app/src/hooks/market-vision/technical-indicators.ts
'use client'

// Basic Moving Averages
export function sma(data: number[], period: number): number[] {
  const result: number[] = new Array(data.length)
  
  // Fill early values with 0 or first available value
  for (let i = 0; i < period - 1; i++) {
    result[i] = 0
  }
  
  for (let i = period - 1; i < data.length; i++) {
    let sum = 0
    let count = 0
    
    for (let j = 0; j < period; j++) {
      const value = data[i - j]
      if (value != null && !Number.isNaN(value) && Number.isFinite(value)) {
        sum += value
        count++
      }
    }
    
    result[i] = count > 0 ? sum / count : 0
  }
  
  return result
}

export function ema(data: number[], period: number): number[] {
  const result: number[] = new Array(data.length)
  const multiplier = 2 / (period + 1)
  
  if (data.length === 0) return result
  
  let firstValidIndex = 0
  while (firstValidIndex < data.length && 
         (data[firstValidIndex] == null || Number.isNaN(data[firstValidIndex] as number) || !Number.isFinite(data[firstValidIndex] as number))) {
    firstValidIndex++
  }
  
  if (firstValidIndex >= data.length) return result.fill(0)
  
  // Fill in values before the first valid index
  for (let i = 0; i < firstValidIndex; i++) {
    result[i] = 0
  }
  
  result[firstValidIndex] = data[firstValidIndex]!
  
  for (let i = firstValidIndex + 1; i < data.length; i++) {
    const value = data[i]
    const prevValue = result[i - 1]
    if (value != null && !Number.isNaN(value) && Number.isFinite(value) && prevValue != null && !Number.isNaN(prevValue) && Number.isFinite(prevValue)) {
      result[i] = (value * multiplier) + (prevValue * (1 - multiplier))
    } else {
      result[i] = prevValue ?? 0
    }
  }
  
  return result
}

export function rma(data: number[], period: number): number[] {
  const result: number[] = new Array(data.length)
  
  if (data.length === 0) return result.fill(0)
  
  // Fill early values with 0
  for (let i = 0; i < period - 1; i++) {
    result[i] = 0
  }
  
  let sum = 0
  let count = 0
  for (let i = 0; i < Math.min(period, data.length); i++) {
    const value = data[i]
    if (value != null && !Number.isNaN(value) && Number.isFinite(value)) {
      sum += value
      count++
    }
  }
  
  if (count > 0) {
    result[period - 1] = sum / count
  } else {
    result[period - 1] = 0
  }
  
  for (let i = period; i < data.length; i++) {
    const value = data[i]
    const prevResult = result[i - 1]
    if (value != null && !Number.isNaN(value) && Number.isFinite(value) && prevResult != null && !Number.isNaN(prevResult) && Number.isFinite(prevResult)) {
      result[i] = (prevResult * (period - 1) + value) / period
    } else {
      result[i] = prevResult ?? 0
    }
  }
  
  return result
}

export function wma(data: number[], period: number): number[] {
  const result: number[] = new Array(data.length)
  
  // Fill early values with 0
  for (let i = 0; i < period - 1; i++) {
    result[i] = 0
  }
  
  for (let i = period - 1; i < data.length; i++) {
    let sum = 0
    let weightSum = 0
    
    for (let j = 0; j < period; j++) {
      const value = data[i - j]
      if (value != null && !Number.isNaN(value) && Number.isFinite(value)) {
        const weight = period - j
        sum += value * weight
        weightSum += weight
      }
    }
    
    result[i] = weightSum > 0 ? sum / weightSum : 0
  }
  
  return result
}

export function hullMA(data: number[], period: number): number[] {
  const halfLength = Math.floor(period / 2)
  const sqrtLength = Math.round(Math.sqrt(period))
  
  const wma1 = wma(data, halfLength)
  const wma2 = wma(data, period)
  
  const diff: number[] = new Array(data.length)
  for (let i = 0; i < data.length; i++) {
    const val1 = wma1[i]
    const val2 = wma2[i]
    if (val1 != null && val2 != null) {
      diff[i] = 2 * val1 - val2
    } else {
      diff[i] = 0
    }
  }
  
  return wma(diff, sqrtLength)
}

export function vwma(prices: number[], volumes: number[], period: number): number[] {
  const result: number[] = new Array(prices.length)
  
  // Fill early values with 0
  for (let i = 0; i < period - 1; i++) {
    result[i] = 0
  }
  
  for (let i = period - 1; i < prices.length; i++) {
    let priceVolumeSum = 0
    let volumeSum = 0
    
    for (let j = 0; j < period; j++) {
      const price = prices[i - j]
      const volume = volumes[i - j]
      if (price != null && volume != null && !Number.isNaN(price) && !Number.isNaN(volume) && Number.isFinite(price) && Number.isFinite(volume)) {
        priceVolumeSum += price * volume
        volumeSum += volume
      }
    }
    
    result[i] = volumeSum > 0 ? priceVolumeSum / volumeSum : 0
  }
  
  return result
}

// RSI Calculation
export function rsi(data: number[], period: number): number[] {
  const result: number[] = new Array(data.length).fill(0) // Initialize with proper length
  const gains: number[] = []
  const losses: number[] = []
  
  // Calculate price changes
  for (let i = 1; i < data.length; i++) {
    const current = data[i]
    const previous = data[i - 1]
    
    if (current != null && previous != null) {
      const change = current - previous
      gains.push(change > 0 ? change : 0)
      losses.push(change < 0 ? Math.abs(change) : 0)
    } else {
      gains.push(0)
      losses.push(0)
    }
  }
  
  const avgGains = rma(gains, period)
  const avgLosses = rma(losses, period)
  
  // Map the RMA results back to the correct indices
  for (let i = 0; i < avgGains.length; i++) {
    const gain = avgGains[i] || 0
    const loss = avgLosses[i] || 0
    
    if (loss === 0) {
      result[i + 1] = 100 // +1 because gains/losses start from index 1
    } else {
      const rs = gain / loss
      result[i + 1] = 100 - (100 / (1 + rs))
    }
  }
  
  return result
}

// Standard Deviation
export function stdev(data: number[], period: number): number[] {
  const result: number[] = new Array(data.length).fill(0) // Initialize with proper length
  
  for (let i = period - 1; i < data.length; i++) {
    const slice = data.slice(i - period + 1, i + 1).filter(val => val != null && !Number.isNaN(val) && Number.isFinite(val))
    
    if (slice.length === 0) {
      result[i] = 0
      continue
    }
    
    const mean = slice.reduce((sum, val) => sum + val, 0) / slice.length
    const variance = slice.reduce((sum, val) => sum + (val - mean) ** 2, 0) / slice.length
    result[i] = Math.sqrt(variance)
  }
  
  return result
}

export function sum(data: number[], period: number): number[] {
  const result: number[] = new Array(data.length)
  
  // Fill early values with 0
  for (let i = 0; i < period - 1; i++) {
    result[i] = 0
  }
  
  for (let i = period - 1; i < data.length; i++) {
    let total = 0
    for (let j = 0; j < period; j++) {
      const value = data[i - j]
      if (value != null && !Number.isNaN(value) && Number.isFinite(value)) {
        total += value
      }
    }
    result[i] = total
  }
  
  return result
}

// ---------------------------------------------------------------------------
// Reverse RSI (Siligardos "Reverse Engineering RSI"), consistent with rsi()
// above: solves for the close the NEXT bar would need for Wilder RSI to print
// a target level. rsi() itself is intentionally untouched.

export interface WilderRsiState {
  avgGain: number
  avgLoss: number
  lastClose: number
}

/**
 * Trailing Wilder avgGain/avgLoss + last close, matching rsi()'s internals
 * exactly (same gains/losses construction, same rma smoothing).
 *
 * Returns null when closes.length < period + 1: below that, the next bar's
 * averages would come from the rma SMA seed row rather than the Wilder
 * recurrence, so the reverse formula would not match the forward rsi().
 */
export function wilderRsiState(closes: number[], period: number): WilderRsiState | null {
  if (!Number.isFinite(period) || period < 1) return null
  if (closes.length < period + 1) return null

  const lastClose = closes[closes.length - 1]
  if (lastClose == null || !Number.isFinite(lastClose)) return null

  // Same gains/losses loop as rsi() so forward and reverse can never drift.
  const gains: number[] = []
  const losses: number[] = []
  for (let i = 1; i < closes.length; i++) {
    const current = closes[i]
    const previous = closes[i - 1]

    if (current != null && previous != null) {
      const change = current - previous
      gains.push(change > 0 ? change : 0)
      losses.push(change < 0 ? Math.abs(change) : 0)
    } else {
      gains.push(0)
      losses.push(0)
    }
  }

  const avgGain = rma(gains, period)[gains.length - 1]
  const avgLoss = rma(losses, period)[losses.length - 1]
  if (avgGain == null || !Number.isFinite(avgGain) || avgLoss == null || !Number.isFinite(avgLoss)) return null

  return { avgGain, avgLoss, lastClose }
}

// The Caretaker's zone levels: critical bull / control bull / scale mid /
// control bear / critical bear.
export const DEFAULT_REVERSE_RSI_TARGETS = [80, 62, 50, 38, 20] as const

export interface ReverseRsiLevel {
  target: number
  price: number | null
}

/**
 * Close price the NEXT bar would need for rsi() to print exactly `target`.
 *
 * With RS = target / (100 - target) and Wilder recurrence avg' = (avg*(p-1) + move)/p:
 * - Up bar   (gain = C - P, valid when RS*avgLoss >= avgGain): C = P + (p-1) * (RS*avgLoss - avgGain)
 * - Down bar (loss = P - C, otherwise):                        C = P - (p-1) * (avgGain/RS - avgLoss)
 *
 * Returns null for unreachable targets (<= 0, >= 100), a flat state
 * (avgGain = avgLoss = 0, where rsi() pins to 100), non-finite inputs, or a
 * solved price <= 0.
 */
export function reverseRsiPrice(state: WilderRsiState, period: number, target: number): number | null {
  if (!Number.isFinite(period) || period < 1) return null
  if (!Number.isFinite(target) || target <= 0 || target >= 100) return null

  const { avgGain, avgLoss, lastClose } = state
  if (!Number.isFinite(avgGain) || !Number.isFinite(avgLoss) || !Number.isFinite(lastClose)) return null
  if (avgGain <= 0 && avgLoss <= 0) return null

  const rs = target / (100 - target)
  const price =
    rs * avgLoss >= avgGain
      ? lastClose + (period - 1) * (rs * avgLoss - avgGain)
      : lastClose - (period - 1) * (avgGain / rs - avgLoss)

  return Number.isFinite(price) && price > 0 ? price : null
}

/**
 * Reverse-RSI price levels for a list of targets (default: Caretaker zones).
 * Prices are null when the state is degenerate or a target is unreachable.
 */
export function reverseRsiLevels(
  closes: number[],
  period: number,
  targets: readonly number[] = DEFAULT_REVERSE_RSI_TARGETS,
): ReverseRsiLevel[] {
  const state = wilderRsiState(closes, period)
  return targets.map((target) => ({
    target,
    price: state ? reverseRsiPrice(state, period, target) : null,
  }))
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Indicators/TechnicalIndicators.swift
import Foundation

/// Port of `hooks/market-vision/technical-indicators.ts` — the LENIENT family (warmup → 0, skips non-finite
/// values). Distinct from `PineMath` (NaN-propagating). Keep both; different indicators rely on each.
public enum TechnicalIndicators {
  @inline(__always) static func ok(_ v: Double) -> Bool { v.isFinite }

  public static func sma(_ data: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: data.count)
    guard period >= 1 else { return result }
    var i = period - 1
    while i < data.count {
      var sum = 0.0, count = 0
      for j in 0..<period { let v = data[i - j]; if ok(v) { sum += v; count += 1 } }
      result[i] = count > 0 ? sum / Double(count) : 0
      i += 1
    }
    return result
  }

  public static func ema(_ data: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: data.count)
    guard !data.isEmpty else { return result }
    let multiplier = 2.0 / Double(period + 1)
    var first = 0
    while first < data.count && !ok(data[first]) { first += 1 }
    guard first < data.count else { return result }
    result[first] = data[first]
    var i = first + 1
    while i < data.count {
      let v = data[i], prev = result[i - 1]
      result[i] = (ok(v) && ok(prev)) ? v * multiplier + prev * (1 - multiplier) : prev
      i += 1
    }
    return result
  }

  public static func rma(_ data: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: data.count)
    guard !data.isEmpty, period >= 1 else { return result }
    var sum = 0.0, count = 0
    for i in 0..<min(period, data.count) { let v = data[i]; if ok(v) { sum += v; count += 1 } }
    if period - 1 < data.count { result[period - 1] = count > 0 ? sum / Double(count) : 0 }
    var i = period
    while i < data.count {
      let v = data[i], prev = result[i - 1]
      result[i] = (ok(v) && ok(prev)) ? (prev * Double(period - 1) + v) / Double(period) : prev
      i += 1
    }
    return result
  }

  public static func wma(_ data: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: data.count)
    guard period >= 1 else { return result }
    var i = period - 1
    while i < data.count {
      var sum = 0.0, weightSum = 0.0
      for j in 0..<period { let v = data[i - j]; if ok(v) { let w = Double(period - j); sum += v * w; weightSum += w } }
      result[i] = weightSum > 0 ? sum / weightSum : 0
      i += 1
    }
    return result
  }

  public static func hullMA(_ data: [Double], _ period: Int) -> [Double] {
    let half = period / 2
    let sqrtLen = Int((Double(period)).squareRoot().rounded())
    let w1 = wma(data, half), w2 = wma(data, period)
    var diff = [Double](repeating: 0, count: data.count)
    for i in 0..<data.count { diff[i] = 2 * w1[i] - w2[i] }
    return wma(diff, sqrtLen)
  }

  public static func vwma(_ prices: [Double], _ volumes: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: prices.count)
    guard period >= 1 else { return result }
    var i = period - 1
    while i < prices.count {
      var pv = 0.0, vs = 0.0
      for j in 0..<period where i - j < volumes.count {
        let p = prices[i - j], v = volumes[i - j]
        if ok(p) && ok(v) { pv += p * v; vs += v }
      }
      result[i] = vs > 0 ? pv / vs : 0
      i += 1
    }
    return result
  }

  public static func rsi(_ data: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: data.count)
    guard data.count > 1 else { return result }
    var gains: [Double] = [], losses: [Double] = []
    gains.reserveCapacity(data.count - 1); losses.reserveCapacity(data.count - 1)
    for i in 1..<data.count {
      let change = data[i] - data[i - 1]
      gains.append(change > 0 ? change : 0)
      losses.append(change < 0 ? -change : 0)
    }
    let avgGains = rma(gains, period), avgLosses = rma(losses, period)
    for i in 0..<avgGains.count {
      let gain = avgGains[i], loss = avgLosses[i]
      result[i + 1] = loss == 0 ? 100 : 100 - 100 / (1 + gain / loss)
    }
    return result
  }

  public static func stdev(_ data: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: data.count)
    guard period >= 1 else { return result }
    var i = period - 1
    while i < data.count {
      let slice = data[(i - period + 1)...i].filter(ok)
      if slice.isEmpty { result[i] = 0; i += 1; continue }
      let mean = slice.reduce(0, +) / Double(slice.count)
      let variance = slice.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / Double(slice.count)
      result[i] = variance.squareRoot()
      i += 1
    }
    return result
  }

  public static func sum(_ data: [Double], _ period: Int) -> [Double] {
    var result = [Double](repeating: 0, count: data.count)
    guard period >= 1 else { return result }
    var i = period - 1
    while i < data.count {
      var total = 0.0
      for j in 0..<period { let v = data[i - j]; if ok(v) { total += v } }
      result[i] = total
      i += 1
    }
    return result
  }

  // MARK: Reverse RSI (Siligardos), consistent with rsi() above.

  public struct WilderRsiState: Sendable, Equatable {
    public var avgGain: Double
    public var avgLoss: Double
    public var lastClose: Double
    public init(avgGain: Double, avgLoss: Double, lastClose: Double) { self.avgGain = avgGain; self.avgLoss = avgLoss; self.lastClose = lastClose }
  }

  public static func wilderRsiState(_ closes: [Double], period: Int) -> WilderRsiState? {
    guard period >= 1, closes.count >= period + 1, let last = closes.last, last.isFinite else { return nil }
    var gains: [Double] = [], losses: [Double] = []
    for i in 1..<closes.count {
      let change = closes[i] - closes[i - 1]
      gains.append(change > 0 ? change : 0)
      losses.append(change < 0 ? -change : 0)
    }
    let g = rma(gains, period)[gains.count - 1], l = rma(losses, period)[losses.count - 1]
    guard g.isFinite, l.isFinite else { return nil }
    return WilderRsiState(avgGain: g, avgLoss: l, lastClose: last)
  }

  public static let defaultReverseRsiTargets: [Double] = [80, 62, 50, 38, 20]

  public struct ReverseRsiLevel: Sendable, Hashable {
    public var target: Double
    public var price: Double?
  }

  public static func reverseRsiPrice(_ state: WilderRsiState, period: Int, target: Double) -> Double? {
    guard period >= 1, target.isFinite, target > 0, target < 100 else { return nil }
    guard state.avgGain.isFinite, state.avgLoss.isFinite, state.lastClose.isFinite else { return nil }
    if state.avgGain <= 0 && state.avgLoss <= 0 { return nil }
    let rs = target / (100 - target)
    let p = Double(period - 1)
    let price = rs * state.avgLoss >= state.avgGain
      ? state.lastClose + p * (rs * state.avgLoss - state.avgGain)
      : state.lastClose - p * (state.avgGain / rs - state.avgLoss)
    return price.isFinite && price > 0 ? price : nil
  }

  public static func reverseRsiLevels(_ closes: [Double], period: Int, targets: [Double] = defaultReverseRsiTargets) -> [ReverseRsiLevel] {
    let state = wilderRsiState(closes, period: period)
    return targets.map { t in ReverseRsiLevel(target: t, price: state.flatMap { reverseRsiPrice($0, period: period, target: t) }) }
  }
}
